### ISSUE TITLE
Upgrade to 2.2.1 binds to php7.1

### ISSUE THREAD

**Issue by Detzler** [magento/magento2#12214]
After updating to 2.2.1 via composer, the package zendframework/zend-code updates to version 3.3.0. This zend-code version requires php not lower than 7.1. It breaks the official system requirements http://devdocs.magento.com/guides/v2.2/install-gde/system-requirements-tech.html

### Preconditions

1. PHP7.1
2. Magento 2.2.1

### Steps to reproduce
1. composer update

### Expected result
1. Magento keeps support for php7.0 in composer

### Actual result
1. Magento allows php not lower than 7.1 via composer

**Comment by orlangur** [comment]
Hi @Detzler, https://github.com/magento/magento2/blob/2.2/composer.json#L11 where did you see PHP 7.0 support dropped? Maybe you did a `composer update` under PHP 7.1?

**Comment by Detzler** [comment]
Sorry, forgot to paste:

```
Problem 1
    - zendframework/zend-code 3.3.0 requires php ^7.1 -> your PHP version (7.0.25) does not satisfy that requirement
```
It appears when installing some other package via php7.0 composer update

> composer update under PHP 7.1

yes, quite possibly

**Comment by orlangur** [comment]
Magento 2.2.1 by itself allows lower version of zend-code: https://github.com/magento/magento2/blob/2.2/composer.json#L13

> yes, quite possibly

Then this report is a duplicate of https://github.com/magento/magento2/issues/11264 and is not a bug. This is how Composer is supposed to work, if you want to get PHP 7.0-compatible packages on any platform, you can fake your platform with https://getcomposer.org/doc/06-config.md#platform

**Comment by czettnersandor** [comment]
The problem is, Magento's composer.json contains this line:

`"zendframework/zend-code": "^3.1.0",`

So `composer update` will fetch the latest version of zend-code, which is currently 3.3.0. This version requires php 7.1. Please understand the difference between tilde (~) and caret (^).

**Comment by czettnersandor** [comment]
I can see it's already fixed here: https://github.com/magento/magento2/commit/42b10d539ca3137563f796eaa30adba6ff8a8fd0

**Comment by orlangur** [comment]
> So `composer update` will fetch the latest version of zend-code, which is currently 3.3.0. This version requires php 7.1

That's exactly how Composer is supposed to work. There is nothing wrong with it.

As far as I remember, Zend Code `3.2.x` supports PHP 7.1+ only and Zend Code `3.1.x` does not support PHP 7.1, so, we cannot use one version for both PHP versions.

I've contacted author of this "fix". If there are known issues with Zend Code `3.3.0` constraint can be changed to `~3.1.0|~3.2.0` but current constraint does not seem to be PHP 7.1 compatible.

**Comment by korostii** [comment]
Hi @orlangur,

>  current constraint does not seem to be PHP 7.1 compatible

It seems to me that there's some issue in there, after all. If that's the case, could you please update the issue status\labels in order to reflect that? Thanks in advance.

**Comment by orlangur** [comment]
@korostii oh, sorry, I mentioned in some other ticket but not this.

I've contacted commit author and we found out that `3.1.x` is PHP 7.1 compatible in fact, just contains less features.

**Comment by korostii** [comment]
Glad to hear.
So, in the end, it's a valid issue which is already fixed in 42b10d5 on branch `2.2-develop` - but not yet released, correct?

**Comment by orlangur** [comment]
@korostii nope, it's non-issue as getting such version of Zend Code when running under PHP 7.1 is expected Composer behavior and is totally fine (assuming Magento is tested with 3.3 of whatever version of package).

Just that 42b10d5 made both PHP 7.0 and 7.1 to resolve the same Zend Code version.
